src/daml/Daml/Trading/Account/Account.daml

```diff
@@ -0,0 +1,211 @@
+module Daml.Trading.Account.Account where 
+
+import DA.Optional (isNone)
+import DA.Map qualified as M (fromList)
+import DA.Set qualified as S (fromList, singleton)
+import DA.List (head, tail)
+import Daml.Finance.Interface.Holding.Factory.Account qualified as Account (Create(..), F)
+import Daml.Finance.Interface.Holding.Factory.Holding qualified as Holding (F)
+import Daml.Finance.Interface.Instrument.Base.Factory qualified as Instrument (F)
+import Daml.Finance.Interface.Types.Common (AccountKey(..), Id(..))
+import Daml.Finance.Interface.Holding.Util (getAmount)
+import Daml.Finance.Interface.Holding.Base qualified as Holding (I)
+import Daml.Finance.Interface.Holding.Fungible qualified as Fungible (I, Split(..), SplitResult(..),Merge(..))
+import Daml.Trading.Reward.Badge qualified as Badge
+import Daml.Trading.Reward.TradingPoint qualified as TradingPoint
+
+cashLabel = "Cash"
+nftLabel = "Nft"
+
+-- to fetch the account factory for subsequent account creation
+template AccountFactoryReference 
+  with 
+    provider: Party 
+    accountFactoryCid: ContractId Account.F 
+  where 
+    signatory provider 
+    key provider: Party 
+    maintainer key 
+
+template HoldingFactoryReference 
+  with 
+    provider: Party 
+    holdingType: Text 
+    holdingFactoryCid: ContractId Holding.F 
+  where 
+    signatory provider 
+    key (provider, holdingType): (Party, Text) 
+    maintainer key._1 
+
+template InstrumentFactoryReference 
+  with 
+    provider: Party 
+    instrumentType: Text 
+    instrumentFactoryCid: ContractId Instrument.F 
+  where 
+    signatory provider 
+    key (provider, instrumentType): (Party, Text) 
+    maintainer key._1 
+
+template Account 
+  with 
+    custodian: Party 
+    owner: Party 
+    accountType: AccountType 
+    username: Text 
+    realName: Text 
+    passportNumber: Text 
+    nationalId: Text 
+    address: Text 
+    optCashAccount: Optional (ContractId Account.F, ContractId Holding.F, AccountKey) 
+    optNftAccount: Optional (ContractId Account.F, ContractId Holding.F, ContractId Instrument.F, AccountKey) 
+    badges : [ContractId Badge.Badge]
+  where 
+    signatory custodian, owner
+    key (custodian, owner): (Party, Party)
+    maintainer key._1 
+
+    choice Update_Permission: ContractId Account 
+      with 
+        newAccountType: AccountType 
+      controller custodian 
+      do 
+        assertMsg "New account type should be different from current one" (accountType /= newAccountType)
+        create this with accountType = newAccountType 
+    
+    choice Create_Cash_Account: ContractId Account 
+      with 
+        description: Text 
+        accountFactoryCid: ContractId Account.F 
+        holdingFactoryCid: ContractId Holding.F 
+        observers: [Party] 
+      controller custodian 
+      do 
+        assertMsg "User already has a cash account" (isNone optCashAccount) 
+
+        let
+          observersSet = S.fromList $ map S.singleton observers
```

Comment:
we pass a list as an argument only to map It to a set of singleton sets? that's weird

Reply:
I am also not sure about this. This pattern is referenced from the quickstart-finance template. It looks like this because the **observers** parameter for creating a daml-finance account has type Map Text (Set (Set Party)). I took a look at the daml-finance library code and it looks like this **observers** Map is flattened into a variable of type Set Party that is used for the template observer